src/channel-qvm.lisp

```diff
@@ -0,0 +1,417 @@
+
+
+(defun requires-swapping-p (qvm)
+  "Does the noisy qvm QVM require swapping of internal pointers?"
+  (and (not (eq (amplitudes qvm) (original-amplitude-pointer qvm)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod run :after ((qvm channel-qvm))
+  ;; Only copy if we really need to.
+  (when (requires-swapping-p qvm)
+    ;; Copy the correct amplitudes into place.
+    (copy-wavefunction (amplitudes qvm) (original-amplitude-pointer qvm))
+    ;; Get the pointer back in home position. We want to swap them,
+    ;; not overwrite, because we want the scratch memory to be intact.
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))))
+
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
+  ;; If any noise rules are matched by theinstruction INSTR, apply the
+  ;; the kraus operators for that noise rule after applying the gate
+  ;; for this instruction.
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar #'(lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (instr-qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (noise-rules (noise-rules (noise-model qvm)))
+         (prepend-kraus-maps (match-rules noise-rules instr :before))
+         (append-kraus-maps (match-rules noise-rules instr :after)))
+    (when prepend-kraus-maps (apply-all-kraus-operators qvm instr prepend-kraus-maps))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple instr-qubits) params)
+    (when append-kraus-maps (apply-all-kraus-operators qvm instr append-kraus-maps))
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod transition :around ((qvm channel-qvm) (instr quil:measurement))
+  (let ((ret-qvm (call-next-method)))
+    (apply-classical-readout-noise ret-qvm instr)
+    ret-qvm))
+
+
+(defun rule-matches-instr-p (rule instr position)
+  "Check if rule is matched by instruction data and the position of the match request."
+  (let* ((noise-predicate (predicate (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr) (string= position noise-position))))
+
+
+(defun match-rules (rules instr position)
+  "Return the operation elements for the first matching rule."
+  (loop :for rule in rules
+        :when (rule-matches-instr-p rule instr position)
+          :return (operation-elements rule)))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-maps)
+  ;; Apply all the kraus operators to the state of the system.
+  (loop :for kraus-map :in kraus-maps
+        :do
+           (check-kraus-ops kraus-map)
+           (apply-kraus-ops qvm instr kraus-map)))
+
+
+(defmethod apply-kraus-ops ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
```

Comment:
Does this need to be a generic function?  It seems to only be called from within `apply-all-kraus-operators`, which has already taken care of specialization on qvm and instruction type.  If possible, consider making it a regular function and renaming to make it clear it's a "helper" for `apply-all-kraus-operators`, like `%apply-one-kraus-operator` or some such.

Reply:
I wanted to have it specialized originally because I am anticipating incorporating different state-representations into the noisy qvms. When this happens, `apply-kraus-ops` will have to specialize on the type of state-representation, as applying kraus ops to a density matrix will be different than applying kraus ops to a pure state. If you think it would be better for me to make this a regular function now and just deal with changing it to a method in a future PR when I actually need to, I can definitely do that.

Reply:
That makes sense. I think it's fine to leave as a generic function if you're pretty sure you'll need it to be in the near future.